### ISSUE TITLE
ThrowableObjectInputStream does not use classloader from settings

### ISSUE THREAD

**Issue by lytvynenkoinvest** [elastic/elasticsearch#12557]
Cluster: 2 nodes
Index not unique document with create flag=true.
node throws DocumentAlreadyExistsException
but this class cannot be loaded  by thread context classloader and
NotSerializableTransportException is thrown.

**Comment by lytvynenkoinvest** [comment]
NettyTransportChannel.sendResponse(Throwable error) line 120 ->
ThrowableObjectOutputStream.canSerialize(error) line 85 ->
ThrowableObjectOutputStream.serialize(T t) line 74 ->
ThrowableObjectInputStream(InputStream in) line 40: constructor does not use classloader from settings

**Comment by lytvynenkoinvest** [comment]
It would be great if classloader will be passed to canSerialize method like this: ThrowableObjectOutputStream.canSerialize(error, transport.settings().getClassLoader())

**Comment by clintongormley** [comment]
Hi @lytvynenkoinvest 

Which version is this in?

**Comment by s1monw** [comment]
Are you running this with assertions enabeld ie. during unittests? can you provide more information about this?

**Comment by lytvynenkoinvest** [comment]
not unit tests.
there are 2 classloaders: 
1) classloader that loads class in which I create and start embedded node, and 
2) threadcontextclassloader that does not contains classes of elasticsearch.
First classloader I set to settings that I put to NodeBuilder.
Then I start 2 nodes in the same cluster, and  I send requests to index(with create=true) in first node.
Periodicaly, first node redirects index requests to second node(round robin).
When second node throws DocumentAlreadyExistsException, the second node checks can this exception be serialized or not. there is the method that serializes exception to byte array and deserializes it and tries to load this class(as I described in issue).
But, to load class it uses threadcontext classloader, not classloader from settings (it is first classloader from list above).
But, ThreadContext classloader does not contain this class and the method throws NotSerializable exception.
